src/Grpc.AspNetCore.Server/Internal/GrpcServiceBinder.cs

```diff
@@ -18,47 +18,98 @@
+        internal IList<IMethod> ServiceMethods { get; } = new List<IMethod>();
 
-        internal GrpcServiceBinder(IEndpointRouteBuilder builder, ServerCallHandlerFactory<TService> serverCallHandlerFactory)
+        internal GrpcServiceBinder(IEndpointRouteBuilder builder, ServiceMethodsRegistery serviceMethodsRegistery, ServerCallHandlerFactory<TService> serverCallHandlerFactory)
         {
             _builder = builder;
+            _serviceMethodsRegistery = serviceMethodsRegistery;
             _serverCallHandlerFactory = serverCallHandlerFactory;
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, ClientStreamingServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateClientStreaming(method);
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, DuplexStreamingServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateDuplexStreaming(method);
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, ServerStreamingServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateServerStreaming(method);
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, UnaryServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateUnary(method);
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
+        }
+
+        private void AddMethodCore(IMethod method, RequestDelegate requestDelegate)
+        {
+            ServiceMethods.Add(method);
+            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, requestDelegate, new object[] { method }));
+        }
+
+        internal void CreateUnimplementedEndpoints()
+        {
+            // Create endpoint that will return UNIMPLEMENTED status for missing method:
+            // - /Package.Service/{method}
+            // Create endpoint that will return UNIMPLEMENTED status for missing service:
+            // - /Package.{service}/{method}
```

Comment:
The case for unimplemented method make sense.

The case for unimplemented service is more tricky and needs some work, see https://github.com/grpc/grpc-dotnet/issues/109#issuecomment-465487229 (perhaps we could leave it for a followup PR?).

Reply:
I'm still not convinced that this way of handling unimplemented services is the correct approach. It's impossible to predict which exact routes will users use in their servers, so it's hard to reason about whether the "/Package.{service}/{method}" route will take precedence over the routes they have and if this will lead to unwanted behavior.

Reply:
Sure. I commented on #109 a few days ago with a solution to also require content type = `application/grpc`. Would adding that be acceptable?

https://github.com/grpc/grpc-dotnet/issues/109#issuecomment-465717408

Reply:
Yes, I think checking the content type is a good solution (and looks like we're planning to do that anyway) See https://github.com/grpc/grpc-dotnet/issues/109#issuecomment-467469913 for details.
Are you planning to update this PR with that approach?